File: web_search.js
'use strict'

//For logging/////
var fs = require('fs');
var util = require('util');
var log_file = fs.createWriteStream(__dirname + '/debug.log', {flags : 'w'});
var log_stdout = process.stdout;

var my_console_log = function(d) { //
	if (util!=null) {
		log_file.write(util.format(d) + '\n');
		log_stdout.write(util.format(d) + '\n');
	}else{
		console.log(d);
	}
};
////////////////////

//Input
var test_input = "Yasaklar konulsa da güneş balçıkla sıvanmaz RTÜK gerekçesi huzur ve güven ortamının bozuyor yasak,Noun konul,Other da,Other güneş,Noun balçık,Noun sıva,Other RTÜK,Noun gerekçe,Noun huzur,Noun ve,Other güven,Noun ortam,Noun boz,Other";


//required global variables
var debug = false;
var https = require('https')
var api_key = 'AIzaSyDoTp6UicPtIH_JVy-cFwoebTEp9-rRHYE';
var api_host = 'kgsearch.googleapis.com';


var main = function (test_input, my_event_emitter) {

	var words = [];
	var roots = [];
	var roots_type = [];
	var printed_results = [];
	var isSomethingFound = false;

	var search_word = function (phrase, callback, limit) {
		if (debug) my_console_log('Searching for phrase :' + phrase);

		if (limit == undefined) {
			limit = 1;
		}

		var query_string = '/v1/entities:search?languages=tr&query=' + phrase + '&key=' + api_key + '&limit=1&indent=True'
		query_string = encodeURI(query_string)
		//my_console_log(query_string)
		https.get({

			host: api_host,
			path: query_string
		}, function (response) {
			// Continuously update stream with data
			var body = '';
			response.on('data', function (d) {
				body += d;
			});
			response.on('end', function () {
				var response = JSON.parse(body);
				var succeed = true;
				// control if it was succeed
				if (debug) my_console_log('Response for phrase :' + phrase + '\n');
				if (debug) my_console_log(response);

				if (response.hasOwnProperty('@context')) {
					succeed = true;

					var itemListElement = response['itemListElement'][0]
					if (debug) my_console_log('ItemlistElement for phrase : ' + phrase);
					if (debug) my_console_log(itemListElement);
					//my_console_log(itemListElement['result']);
					///my_console_log(itemListElement['resultScore']);
					if (itemListElement != [] && itemListElement != undefined)  {
						if(debug) my_console_log('Callback for phrase : '+ phrase);
						callback(succeed, itemListElement['result'], itemListElement['resultScore'])
					} else {
						if (debug) my_console_log('Failed callback for phras : '+ phrase + '\n');
						callback(true, null, null)
					}
				}else{
					if (debug) my_console_log('Failed result for phrase : '+ phrase);
					callback(false,null, null)
				}


				//my_console_log(response)
			});
		});
	}

//testing
// search_word('google', function(result, response){});

	var init_state_machine = function (input) {
		if (debug) my_console_log('State machine is initialized with input: ')
		if (debug) my_console_log(input);

		var input_array = input.split(" ");
		words = input_array.slice(0, input_array.length / 2);
		var temp_roots = input_array.slice(input_array.length / 2, input_array.length);

		for (var i = 0; i < temp_roots.length; i++) {
			var root_and_type = temp_roots[i].split(',');

			roots_type.push(root_and_type[1]);
		    roots.push(root_and_type[0]);
		}
		if(debug) my_console_log('Input words : ');
		if(debug) my_console_log(words);
		if(debug) my_console_log('Input roots : ');
		if(debug) my_console_log(roots);

		first_search(0);
	}

	var first_search = function (index) {

		search_word(words[index], function (isSucceed, result, score) {

			if (isSucceed) {
				if (debug) my_console_log('Callback for word is succeed :' + words[index]);
				if (score != null) {
					if (index + 1 < words.length) {
						if (debug) my_console_log('Word is going to iteration : ' + words[index]);
						iterasyon(result, score, index + 1, words[index])
					} else {
						if (roots_type[index]=='Noun') {
							if(debug) my_console_log('last word is noun here is the result of :' + words[index])
							if (debug) my_console_log(result);
							printed_results.push(result);
							isSomethingFound = true
							if(my_event_emitter!=null) my_event_emitter.emit('result',result)
						} else {
							if(debug) my_console_log('last word is not noun :' + words[index])
							if (debug) my_console_log(result);
							//save for causion
						}
						finish();
					}
				} else {
					if (debug) my_console_log('Word is going to root_serach : ' + words[index]);
					root_search(index)
				}
			}else{
				if (debug) my_console_log('Callback for word is failed :' + words[index]);
			}
		})
	}

	var root_search = function (index) {
		search_word(roots[index], function (isSucceed, result, score) {
			if (isSucceed) {
				if (debug) my_console_log('Callback for root is succeed :' + roots[index]);
				if (score != null) {
					if (index + 1 < words.length) {
						if (debug) my_console_log('Root is going to iteration : ' + roots[index]);
						iterasyon(result, score, index + 1, roots[index])
					} else {
						if (roots_type[index]=='Noun') {
							if(debug) my_console_log('Last root is noun here is the result of :' + roots[index])
							if (debug) my_console_log(result);
							printed_results.push(result);
							isSomethingFound = true
							if(my_event_emitter!=null) my_event_emitter.emit('result',result);
						} else {
							if (debug) my_console_log('Last root is not noun :' + roots[index])
							if (debug) my_console_log(result);
							//save for causion
						}
					}
				} else {
					if (debug) my_console_log('Zero score even in the root. It is going next word search :' + roots[index])
					if (debug) my_console_log(result);
					first_search(index + 1)
				}
			}else{
				if (debug) my_console_log('Callback for root is failed :' + roots[index]);
			}
		})
	}
	var iterasyon = function (result, score, index, phrase) {
		var new_phrase = phrase + ' ' + words[index];
		if (debug) my_console_log('Iteration for phrase : ' + new_phrase);

		search_word(new_phrase, function (isSucceed, result2, score2) {
			if (isSucceed) {
				if (debug) my_console_log('Callback for iteration phrase is succeed :' + new_phrase);
				if (debug) my_console_log('Score for phrase is' + score2  + ' : ' + new_phrase)
				if (score2 != null && score2 >= score) {
					if (debug) my_console_log('iteration phrase is going next iteration :' + new_phrase);
					iterasyon(result2, score2, index + 1, new_phrase)
				} else if (score2 == null || score2 < score) {
					if (debug) my_console_log('iteration phrase is going root iteration :' + new_phrase);
					root_iterasyon(result, score, index, phrase)
				}
			}else{
				if (debug) my_console_log('Callback for iteration phrase is failed :' + new_phrase);
			}
		})
	}

	var root_iterasyon = function (result, score, index, phrase) {
		var new_phrase = phrase + ' ' + roots[index];
		if (debug) my_console_log('Root iteration for phrase : ' + new_phrase);
		search_word(new_phrase, function (isSucceed, result2, score2) {
			if (isSucceed) {
				if (debug) my_console_log('Callback for root iteration phrase is succeed :' + new_phrase);
				if (debug) my_console_log('Score for phrase is' + score2 + ' : ' + new_phrase)
				if (score2 != null && score2 >= score) {
					if (debug) my_console_log('Root iteration phrase is going next iteration :' + new_phrase);
					iterasyon(result2, score2, index + 1, new_phrase)
				} else if (score2 == null || score2 < score) {
					if (debug) my_console_log('Root iteration phrase is going to matching :' + new_phrase);
					matching(result, score, index, phrase)
				}
			}else{
				if (debug) my_console_log('Callback for root iteration phrase is failed :' + new_phrase);
			}
		})
	}

	var matching = function (result, score, index, phrase) {
		if (debug) my_console_log('Matching result for phrase : ' + phrase + '\n');
		if (debug) my_console_log(result)

		var words_unreached = words.slice(index, words.length)
		if (result.name!=undefined) {
			var result_unreached = result.name.toLowerCase().replace(phrase, '').split(" ").splice(1)
		}else{
			var result_unreached = result.description.toLowerCase().replace(phrase, '').split(" ").splice(1)
		}

		var found = false;

		var last_index = index;

		for (var i = 0; i < result_unreached.length; i++) {

			if (i < words_unreached.length) {

				if (words_unreached[i].length > result_unreached[i].length) {
					if (words_unreached[i].indexOf(result_unreached[i]) != -1) {
						index = index + 1
						found = true
					} else {
						break;

					}
				} else {
					if (result_unreached[i].indexOf(words_unreached[i]) != -1) {
						found = true;
						index = index + 1
					} else {
						break;
					}
				}
			} else {
				break;
			}

		}

		if (phrase.split(" ").length == 1 && roots_type[last_index] == 'Noun') {
			if (debug) my_console_log('Result for phrase will be printed : ' + phrase);
			printed_results.push(result);
			isSomethingFound = true
			if(my_event_emitter!=null) my_event_emitter.emit('result',result)

		} else if(phrase.split(" ").length  > 1){
			if (debug) my_console_log('Result for phrase will be printed : ' + phrase);
			printed_results.push(result);
			isSomethingFound = true
			if(my_event_emitter!=null) my_event_emitter.emit('result',result)
			//print that results
		}else{
			if (debug) my_console_log('Result for phrase will not be printed because it is not noun : ' + phrase);
		}

		if (index < words.length) {
			if (debug) my_console_log('System goes new search ');
			first_search(index);
		} else {
			if(debug) my_console_log('System will be shutdown')
			finish()
		}
	}


	var finish = function () {
		if (debug) my_console_log('finish')
		if (!isSomethingFound) {
		if (debug)	my_console_log('Nothing found');
			if (event_emitter != null){
				event_emitter.emit("error", 'Nothing found')
			}
		}
		if (debug) my_console_log('-----------------Printed Results------------------------');
		else console.log('-----------------Printed Results------------------------');
		for(var i = 0 ; i < printed_results.length; i++){
			if (printed_results[i].name != undefined) {
				var name = printed_results[i].name
			}else{
				var name = printed_results[i].description
			}
			if (debug) my_console_log(name);
			else console.log(name);
		}

		//evaluate if nothing is printed
	}


	init_state_machine(test_input);
}

//Event usage
const event_emitter = require('events').EventEmitter;
const myEmitter = new event_emitter();

main(test_input, myEmitter);

myEmitter.on('result', function(result){
	console.log('result: ' + result);
});

File: src/WebSearch.py
import requests, json
from requests.utils import quote

def my_console_log(phrase) :
    print phrase


def console_log(phrase) :
    print phrase


class WebSearch:
    api_host = "https://kgsearch.googleapis.com";

    def __init__(self, api_key="AIzaSyDoTp6UicPtIH_JVy-cFwoebTEp9-rRHYE",debug=False) :
        self.api_key = api_key
        self.words = []
        self.roots = []
        self.roots_type = []
        self.printed_results = []
        self.isSomethingFound = False
        self.debug = debug
        self.my_event_emitter = None

    def starter(self, test_input, debug, my_event_emitter, my_console_log) :
        self.debug = debug
        self.my_event_emitter = my_event_emitter
        if len(test_input) > 2:
             self.init_state_machine(test_input)
             return self.printed_results
        else:
            return [""]

    def search_word(self, phrase, limit=1) :
        if self.debug : my_console_log("Searching for phrase :" + phrase)
        query_string = "/v1/entities:search?languages=tr&query=" + quote(phrase.encode("utf-8")) + "&key=" + self.api_key + "&limit=1&indent=True"
        query_string = WebSearch.api_host+query_string
        r = requests.get(query_string)
        response = r.json()
        succeed = r.status_code == 200
        if self.debug: my_console_log("Response for phrase :" + phrase + "\n")
        if self.debug: my_console_log(response)
        if '@context' in response:
            succeed = True
            itemListElement = response["itemListElement"]
            if len(itemListElement ) > 0:
                itemListElement = itemListElement[0]
                if self.debug : my_console_log("ItemlistElement for phrase : " + phrase)
                if self.debug : my_console_log(itemListElement)
                if self.debug : my_console_log(itemListElement["result"])
                if self.debug : my_console_log(itemListElement["resultScore"])
            else:
                itemListElement = None


            if itemListElement != [] and itemListElement != None :
                if self.debug : my_console_log("Callback for phrase : "+ phrase)
                return succeed, itemListElement["result"], itemListElement["resultScore"]
            else :
                if self.debug : my_console_log("Failed callback for phras : "+ phrase + "\n")
                return True, None, None;

        else:
            if self.debug : my_console_log("Failed result for phrase : "+ phrase)
            return False, None, None;

    def first_search(self, index) :
        isSucceed, result, score = self.search_word(self.words[index])
        if isSucceed:
            if self.debug: my_console_log("Callback for word is succeed :" + self.words[index]);
            if score is not None:
                if index + 1 < len(self.words) :
                    if self.debug: my_console_log("Word is going to iteration : " + self.words[index]);
                    return self.iterasyon(result, score, index + 1, self.words[index])
                else:
                    if self.roots_type[index] == "Noun" :
                        if self.debug: my_console_log("last word is noun here is the result of :" + self.words[index]);
                        if self.debug: my_console_log(result);
                        self.printed_results.append(result)
                        self.isSomethingFound = True
                        if self.my_event_emitter is not None: my_event_emitter.emit("result",result);
                    else:
                        if self.debug: my_console_log("last word is not noun :" + self.words[index]);
                        if self.debug: my_console_log(result);
                        #save for causion
                    print "--firstsearch calls finish"
                    return self.finish()
            else :
                if self.debug : my_console_log("Word is going to root_serach : " + self.words[index])
                return self.root_search(index)
        else:
            if self.debug : my_console_log("Callback for word is failed :" + self.words[index])

    def root_search(self, index):
        isSucceed, result, score = self.search_word(self.words[index])
        if isSucceed:
            if self.debug: my_console_log("Callback for root is succeed :" + self.roots[index])
            if score is not None:
                if index + 1 < len(self.words):
                    if self.debug : my_console_log("Root is going to iteration : " + self.roots[index])
                    return self.iterasyon(result, score, index + 1, self.roots[index])
                else:
                    if self.roots_type[index] == "Noun":
                        if self.debug: my_console_log("Last root is noun here is the result of :" + self.roots[index])
                        if self.debug: my_console_log(result)
                        self.printed_results.append(result)
                        self.isSomethingFound = True
                        if my_event_emitter is not None:
                            my_event_emitter.emit("result", result)
                    else:
                        if self.debug: my_console_log("Last root is not noun :" + self.roots[index])
                        if self.debug: my_console_log(result);
                        #save for causion
                    print "--rootsearch calls finish1"
                    return self.finish()
            else:
                if self.debug: my_console_log("Zero score even in the root. It is going next word search :" + self.roots[index])
                if self.debug: my_console_log(result)
                if index + 1 < len(self.words):
                    return self.first_search(index + 1)
                else:
                    print "--rootsearch calls finish2"
                    return self.finish()

        else:
            if self.debug : my_console_log("Callback for root is failed :" + self.roots[index]);
            return ""

    def iterasyon(self, result, score, index, phrase) :
        new_phrase = phrase + " " + self.words[index]
        if self.debug:
            my_console_log("Iteration for phrase : " + new_phrase);

        isSucceed, result2, score2 = self.search_word(self.words[index])
        if isSucceed :
            if self.debug: my_console_log("Callback for iteration phrase is succeed :" + new_phrase);
            if score2 is not None and score2 >= score :
                if self.debug: my_console_log("Score for phrase is" + str(score2) + " : " + new_phrase);
                if index + 1 < len(self.words) :
                    if self.debug: my_console_log("iteration phrase is going next iteration :" + new_phrase);
                    return self.iterasyon(result2, score2, index + 1, new_phrase);
                else:
                    if self.debug: my_console_log("Last iteration the result of :" + new_phrase);
                    if self.debug: my_console_log(result2);
                    self.printed_results.append(result2)
                    self.isSomethingFound = True;
                    if self.my_event_emitter is not None :
                        self.my_event_emitter.emit("result",result2);
                    return self.finish()
            elif score2 is None or score2 < score:
                if self.debug: my_console_log("iteration phrase is going root iteration :" + new_phrase)
                return self.root_iterasyon(result, score, index, phrase)
        else :
            if self.debug : my_console_log("Callback for iteration phrase is failed :" + new_phrase);
            return ""

    def root_iterasyon(self,result, score, index, phrase) :

        new_phrase = phrase + " " + self.roots[index]

        if self.debug:
            my_console_log("Root iteration for phrase : " + new_phrase);
        isSucceed, results, score2 = self.search_word(self.words[index]);
        if isSucceed:
            if self.debug: my_console_log("Callback for root iteration phrase is succeed :" + new_phrase);
            if score2 is not None and score2 >= score :
                if self.debug : my_console_log("Score for phrase is" + str(score2) + " : " + new_phrase)
                if index + 1 < self.words.length :
                    if self.debug: my_console_log("Root iteration phrase is going next iteration :" + new_phrase);
                    return self.iterasyon(result2, score2, index + 1, new_phrase)
                else :
                    if self.debug: my_console_log("Last root iteration the result of :" + new_phrase)
                    if self.debug: my_console_log(result2);
                    self.printed_results.append(result2)
                    self.isSomethingFound = true
                    if my_event_emitter is not None:
                        my_event_emitter.emit("result", result2)
                    return self.finish()

            elif score2 is None or score2 < score:
                if self.debug :
                    my_console_log("Root iteration phrase is going to matching :" + new_phrase)
                return self.matching(result, score, index, phrase)
        else:
            if self.debug: my_console_log("Callback for root iteration phrase is failed :" + new_phrase)

    def init_state_machine(self, input):
        if self.debug :
            my_console_log("State machine is initialized with input: ")
        if self.debug :
            my_console_log(input)
        if input != "" :
            input_array = input.split(" ")
            print input_array
            self.words = input_array[slice(0, len(input_array) / 2)]
            temp_roots = input_array[slice(len(input_array) / 2, len(input_array))];
            for i in range(0, len(temp_roots)):
                root_and_type = temp_roots[i].split(",")
                self.roots_type.append(root_and_type[1])
                self.roots.append(root_and_type[0])
            if self.debug: my_console_log("Input words : ")
            if self.debug: my_console_log(self.words);
            if self.debug: my_console_log("Input roots : ")
            if self.debug: my_console_log(self.roots);
            return self.first_search(0)
        else:
            print "init state machine calls finish"
            return self.finish()

    def matching(self, result, score, index, phrase):
        if self.debug:
            my_console_log("Matching result for phrase : " + phrase + "\n")
        if self.debug:
            my_console_log(result)

        words_unreached = self.words[slice(index, len(self.words))]
        result_unreached = None
        if 'name' in result:
            result_unreached = result['name'].lower().replace(phrase, "").split(" ")[1:]
        elif 'description' in result:
            result_unreached = result['description'].lower().replace(phrase, "").split(" ")[1:]

        found = False
        last_index = index

        if result_unreached is not None:
            for i in range(0, len(result_unreached)):
                if i < len(words_unreached):
                    if len(words_unreached[i]) > len(result_unreached[i]) :
                        try:
                            words_unreached[i].index(result_unreached[i])
                            index = index + 1
                            found = True
                        except ValueError:
                            break
                    else:
                        try:
                            result_unreached[i].index(words_unreached[i])
                            found = True
                            index = index + 1
                        except ValueError:
                            break
                else:
                    break
        if 'name' not in result:
            result['name'] = phrase

        if len(phrase.split(" ")) == 1 and self.roots_type[last_index] == "Noun":
            if self.debug: my_console_log("Result for phrase will be printed : " + phrase)
            self.printed_results.append(result)
            self.isSomethingFound = True
            if self.my_event_emitter is not None : my_event_emitter.emit("result",result)
        elif len(phrase.split(" ")) > 1:
            if self.debug: my_console_log("Result for phrase will be printed : " + phrase)
            self.printed_results.append(result)
            self.isSomethingFound = True
            if self.my_event_emitter is not None : my_event_emitter.emit("result",result)
            #print that results
        else :
            if self.debug : my_console_log("Result for phrase will not be printed because it is not noun : " + phrase)

        if index < len(self.words):
            if self.debug: my_console_log("System goes new search ")
            return self.first_search(index)
        else:
            if self.debug:  my_console_log("System will be shutdown")
            return self.finish()

    def finish(self):

        if self.debug:
            my_console_log("finish")

        if not self.isSomethingFound:
            if self.debug:
                my_console_log("Nothing found")
            if self.my_event_emitter is not None:
                my_event_emitter.emit("error_occurs", "NothingFound")
        else:
            if self.my_event_emitter is not None:
                my_event_emitter.emit("finished", "SearchIsFinished")

        if self.debug:
            my_console_log("-----------------Printed Results------------------------");
        else:
            console_log("-----------------Printed Results------------------------");

        for i in range(0, len(self.printed_results)):
            if 'name' in self.printed_results[i]:
                name = self.printed_results[i]['name']
            else:
                name = printed_results[i]['description']

            if self.debug:
                my_console_log(name);
            else:
                console_log(name);
        return self.printed_results
        #console.timeEnd("websearch")
